packages/@react-stately/selection/src/useMultipleSelectionState.ts

```diff
@@ -110,7 +110,8 @@ export function useMultipleSelectionState(props: MultipleSelectionStateProps): M
     },
     selectedKeys,
     setSelectedKeys(keys) {
-      if (allowDuplicateSelectionEvents || !equalSets(keys, selectedKeys)) {
+      let isKeysMatched = keys === 'all' && selectedKeys === 'all';
+      if (isKeysMatched && allowDuplicateSelectionEvents || !equalSets(keys, selectedKeys)) {
```

Comment:
I think you'll want something like this
```suggestion
      let isBothAll = keys === 'all' && selectedKeys === 'all';
      if (allowDuplicateSelectionEvents || (!isBothAll && !equalSets(keys, selectedKeys)))   
```

so that `equalSets` doesn't get called if both are 'all'.

After taking another look, it might actually be a bit easier to just update `equalSets` to handle 'all' being passed in. It could be renamed to `equalSelections`:

```ts
function equalSelections(setA, setB) {
  if (setA === 'all' && setB === 'all') {
    return true
  } else if (setA === 'all' || setB === 'all') {
    return false.
  }

  if (setA.size !== setB.size) {
    return false;
  }

  for (let item of setA) {
    if (!setB.has(item)) {
      return false;
    }
  }

  return true;
}
```

Either of those should work. The second feels a bit cleaner and that function is only used there, so it isn't a huge change.

Reply:
That should work too! Feel free to do that instead.